### PR TITLE
Added paged retrieval approach to get larger BMs

### PR DESCRIPTION
**Pull request by rohitsngh27**
# Resolves
https://jira.rax.io/browse/SALUS-1031

# What
Added paged retrieval approach to get larger BMs

# How
Used paged traversal for retrieving large BMs and delete them.

# How to test
This can be tested via Junit and Rest call.

# Why
This approach seems to resolve the OOM error issue and speed up the deleting process too.

### CODE REVIEW

File: src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementPolicyTest.java [hunk 1]
```diff
@@ -104,6 +104,7 @@
 import org.springframework.data.domain.PageImpl;
 import org.springframework.data.domain.PageRequest;
 import org.springframework.data.domain.Pageable;
+import org.springframework.data.domain.Range.Bound;
```
Comment (jjbuchan) [inline src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementPolicyTest.java:107]: This is unused.

File: src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java [hunk 2]
```diff
@@ -1528,11 +1526,12 @@ public void testRemoveMonitor() {
...
-
-    when(boundMonitorRepository.findAllByTenantIdAndMonitor_IdIn(anyString(), any()))
-        .thenReturn(Collections.singletonList(boundMonitor));
-
+    BoundMonitor b = new BoundMonitor();
+    b.setMonitor(monitor);
+    b.setResourceId("r-1");
+    b.setEnvoyId("e-goner");
+    when(boundMonitorRepository.findAllByTenantIdAndMonitor_IdIn(anyString(), any(), any()))
+        .thenReturn(new PageImpl<>(Collections.singletonList(b))).thenReturn(Page.empty());
```
Comment (jjbuchan) [inline src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java:1534]:
```suggestion
    when(boundMonitorRepository.findAllByTenantIdAndMonitor_IdIn(anyString(), any(), any()))
        .thenReturn(new PageImpl<>(Collections.singletonList(b)))
        .thenReturn(Page.empty());
```

Can you switch it to this so its a little clearer that multiple returns are being used.

**Review by jjbuchan — CHANGES_REQUESTED** [review]
A couple of minor points.

File: src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java [hunk 3]
```diff
@@ -1528,13 +1526,16 @@ public void testRemoveMonitor() {
...
 
-    final BoundMonitorProjection boundMonitor = new BoundMonitorProjection("e-goner", monitor, "t-1", "r-1", "zoneA");
-
-    when(boundMonitorRepository.findAllByTenantIdAndMonitor_IdIn(anyString(), any()))
-        .thenReturn(Collections.singletonList(boundMonitor));
-
+    BoundMonitor b = new BoundMonitor();
+    b.setMonitor(monitor);
+    b.setResourceId("r-1");
+    b.setEnvoyId("e-goner");
```
Comment (itzg) [inline src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java:1532]: No need to change, but FYI we have lombok's chainable setters enabled, so you can save a few characters using:
```suggestion
    BoundMonitor b = new BoundMonitor()
      .setMonitor(monitor)
      .setResourceId("r-1")
      .setEnvoyId("e-goner");
```

File: src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java [hunk 4]
```diff
@@ -3268,11 +3271,18 @@ public void deleteAllMonitorsForTenant() {
...
+
     monitorManagement.removeAllTenantMonitors("t-1", false);
 
     Page<Monitor> result = monitorManagement.getMonitors("t-1", Pageable.unpaged());
+    List<UUID> monitorIds = result.get()
+        .map(Monitor::getId)
+        .collect(Collectors.toList());
 
-    verify(boundMonitorRepository).findAllByTenantIdAndMonitor_IdIn(any(), any());
+    verify(boundMonitorRepository).findAllByTenantIdAndMonitor_IdIn(any(), any(), any());
```
Comment (itzg) [inline src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java:3285]: I guess we missed this in the original PR for this code. Can you try changing these to exact argument matchers? That might be why you added the `monitorIds` extraction above?

Comment (rohitsngh27) [reply src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java:3285]: Ohh, Got it now

File: src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java [hunk 5]
```diff
@@ -3268,11 +3271,18 @@ public void deleteAllMonitorsForTenant() {
...
     createMonitorsForTenant(20, "t-1");
 
+    when(boundMonitorRepository
+        .findAllByTenantIdAndMonitor_IdIn(any(), any(), any()))
+        .thenReturn(Page.empty());
+
     monitorManagement.removeAllTenantMonitors("t-1", false);
 
     Page<Monitor> result = monitorManagement.getMonitors("t-1", Pageable.unpaged());
+    List<UUID> monitorIds = result.get()
```
Comment (itzg) [inline src/test/java/com/rackspace/salus/monitor_management/services/MonitorManagementTest.java:3281]: `monitorIds` doesn't seem to be used. (That's what I meant in comment below.)

**Review by jjbuchan — APPROVED** [review]

**Pull request merged**